### ISSUE TITLE
Reading file unlocks it?

### ISSUE THREAD

**Issue by LuKks** [mafintosh/fsctl#5]
You can run in two terminals `node lock.js` and the lock will be granted to both procceses.
Later, comment the line with `fs.readFileSync` and try again, lock will only be granted to one process.

Is this behaviour expected? I guess I will start using a separated file just for locking purposes to avoid breaking the lock.

`lock.js`
```js
const fs = require('fs')
const { open } = require('fs/promises')
const { tryLock, unlock } = require('fsctl')

main()

async function main () {
  const file = await open('file.txt', 'a+')

  const granted = tryLock(file.fd, { exclusive: true })
  console.log('lock granted?', granted)

  if (!granted) {
    return
  }

  console.log('reading', fs.readFileSync('file.txt', 'utf8'))

  console.log('holding lock 15s')
  await sleep(15000)

  unlock(file.fd)
}

function sleep (ms) {
  return new Promise(r => setTimeout(r, ms))
}
```

**Comment by mafintosh** [comment]
Let me try myself, but in general I'd say a much better pattern is to use the file descriptor to issue reads after (what hypercore does)

**Comment by mafintosh** [comment]
As in, when using a locked fd, you should always use a locked fd. readFileSync will open its own fd without a lock

**Comment by mafintosh** [comment]
Actually I think something weird is going on, let me investigate

**Comment by mafintosh** [comment]
Ah no, had a typo. Here is how you should do it:

```js
async function main () {
  const file = await open('file.txt', 'a+')

  const granted = tryLock(file.fd, { exclusive: true })
  console.log('lock granted?', granted)

  if (!granted) {
    return
  }

  for await (const data of file.createReadStream({ autoClose: false })) {
    console.log(data)
  }

  console.log('holding lock 15s')
  await sleep(15000)

  unlock(file.fd)
}
```

**Comment by LuKks** [comment]
> As in, when using a locked fd, you should always use a locked fd. readFileSync will open its own fd without a lock

The idea of using `fs.readFileSync` in the example was to show the worst case scenario, ie. imagine wanting to lock a file and later it gets written by another library which open/closes the fd.
With your example, now I understand that it's about not closing the fd, or about not opening a new fd which will ignore the lock.

My case it's a multiple JSON files that gets read/written several times across multiple procceses, so I have control over the fd, thanks!

Btw, this library is awesome! Kasper seems godlike with his crazy pull requests!

**Comment by LuKks** [comment]
Wow, same code works now, using latest commit and Node 16.
Previously, I was using `mutexify` and some more code to avoid "locking" the same file in the same process.
Now it's much simpler! I will continue using `mutexify` to avoid a case of opening too many fd in a single process, still the code is simpler.

**Comment by mafintosh** [comment]
@LuKks awesome!

@kasperisager We should the initial snippet as a test case. Ie open and closing the same file unlocked, doesn’t affect the lock

**Comment by kasperisager** [comment]
@LuKks @mafintosh This was fixed with the introduction of BSD locks in #6. I'll add a test case for if/when we reintroduce POSIX based locks 👍

Edit: Ready in #8.
